### ISSUE TITLE
🚀 Feature: Delete all user sessions on blocking a user

### ISSUE THREAD

**Issue by prechILLEd** [appwrite/appwrite#6061]
### 🔖 Feature description

When deleting a user in the appwrite backend the sessions are still stored and not deleted. 

Therefore a different user using the same device wont be able to sign in because the request is completely blocked.

### 🎤 Pitch

In my use-case it would make more sense to delete all sessions when a user is blocked since other users might not be able to sign in because the requests get rejected due to the blocked users sessions.

Steps:
1. Sign in with user A
2. Block user A in the appwrite backend
3. Restart App/Reload Website
4. Try to sign in with User B
5. Getting "AppwriteException: user_blocked, The current user has been blocked. You can unblock the user from the Appwrite console. (401)"

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/.github/blob/main/CODE_OF_CONDUCT.md)

**Comment by JorensM** [comment]
I think an even better solution would be to automatically end the previous session if a new session is being created. I'm not sure about the security concerns of this though

**Comment by prechILLEd** [comment]
> 

For that use-case it is probably best to set the session limit per user to 1 in the Auth Security tab.

You might want a user to be logged in on multiple devices at the same time and your thought would deny that.

**Comment by joeyouss** [comment]
Hi, thank you for opening this issue, but would it make sense for you to maybe set the limit once and then try it out? https://appwrite.io/docs/authentication-security#limits

I thinj, there are some potential downsides to doing this. In some cases, retaining the session might be necessary for auditing or investigative purposes, especially in applications that require high security. Also, automatically deleting sessions could result in data loss if the user was in the middle of an unsaved task. There might be scenarios where an account is temporarily blocked due to some system or admin error. In such cases, deleting all sessions would force the user to log back into all devices, which could be an inconvenience.

Can you let me know if setting the limit does the work for you and what are your thoughts?

**Comment by prechILLEd** [comment]
Thanks @joeyouss for the detailed answer.

You are totally right. You do not want to delete all of the users sessions for that purposes.
For my in-house app it does not really matter if data is lost.

Even with a session limit set to 1 my issue would still be not fixed - beside the fact it is not my goal to limit the user to have only 1 active session.

As a workaround I wrote a function that deletes all sessions for every user that is blocked. 
It gets triggered when a user tries to sign in and  get a "AppwriteException: user_blocked, The current user has been blocked. You can unblock the user from the Appwrite console. (401)" and tries to sign in again with the credentials given. This works fine for me!

I think it is not necessary to still work on this issue due to Joe's explanation.

**Comment by prechILLEd** [comment]
Just noticed I can not execute the function from that session since the "current user is blocked" and can not execute a function with "any" execute access.

Any workaround to recommend?

**Comment by prechILLEd** [comment]
Changed to a scheduled execution every 15 minutes which is totally fine for me.

For me this is kinda solved now with this workaround.

**Comment by sudormrfbin** [comment]
> Changed to a scheduled execution every 15 minutes which is totally fine for me.

Another option would be to write a function that is triggered on the `users.*.update.status` event.

There is a small caveat tho, as mentioned in the [Known Limitations][1] section:

> When events fire, only existing subscriptions for that event will receive the update. If your client or server side integrations lose network connection temporarily, delivery of the event is not guaranteed.
>
> For self-hosted instances, when the Appwrite containers are shut down and restarted, events with pending webhooks and subscription updates will not be delivered.

[1]: https://appwrite.io/docs/events#events-limitations

**Comment by stnguyen90** [comment]
We would like to move forward with updating the [Update user status](https://github.com/appwrite/appwrite/blob/45c44bb51c7438010b3649cf177aae2e57f790f8/app/controllers/api/users.php#L698-L699) API to also delete sessions.

**Comment by SophieNguyen113** [comment]
**We're working on this**

**Team:** CodeDay Labs 2024
**Mentor:** Robinson Ibarra
**Mentees:** [Sophie Nguyen](https://github.com/SophieNguyen113), Nha Pham, Rojee Koju, Edale Miguel

**Comment by stnguyen90** [comment]
@SophieNguyen113, how's your progress on this? FYI, I'll need to unassign you soon if there's been no activity.

**Comment by SophieNguyen113** [comment]
We're working on this. The deadline is the end of next week. Thanks for
understanding.

**Comment by raibarra04** [comment]
@stnguyen90 Hello Steven, we have the project set up locally and have some plans on how to update the User Status API. 

Once we start testing the feature, what would be the best way to test with Postman? Or would you mind pointing us in the right direction to finding this information?

**Comment by stnguyen90** [comment]
@raibarra04, there should be a E2E test case added for this.

**Comment by SophieNguyen113** [comment]
Hello, I've opened a [PR](https://github.com/appwrite/appwrite/pull/8545) to update the User Status API to delete all sessions when a user is blocked. Please review!
@stnguyen90  @raibarra04

**Comment by SophieNguyen113** [comment]
I saw some failed tests. I'll fix it soon. @stnguyen90 @raibarra04

**Comment by raibarra04** [comment]
@stnguyen90 Just a small update on this. 

Unfortunately, Sophie left the team, so we adjusted the changes she had made and will be submitting a new PR soon with the code changes and a new E2E test.

**Comment by SophieNguyen113** [comment]
> @stnguyen90 Just a small update on this.
> 
> Unfortunately, Sophie left the team, so we adjusted the changes she had made and will be submitting a new PR soon with the code changes and a new E2E test.

I'm so sorry for slacking off recently. I managed to solve the issue, and my [Pull Request](https://github.com/appwrite/appwrite/pull/8545) is ready to review. Thanks.

**Comment by SophieNguyen113** [comment]
@stnguyen90  

This isn't a big problem enough to open an issue I think. I don't know if anyone is working on this, but if there is no one yet, I would like to open a new PR for [these](https://github.com/appwrite/appwrite/blob/main/tests/e2e/Services/Users/UsersBase.php#L1579-L1580).

**Comment by SophieNguyen113** [comment]
@stnguyen90  @raibarra04  Hello, since I accidentally messed up my old PR, I would close it and open this new [PR](https://github.com/appwrite/appwrite/pull/8660). This new one should be the most updated. Please review. Thanks.

**Comment by EJM143** [comment]
@stnguyen90 Hi Steven, we have submitted a new PR for the feature. Please review our changes when you have time and let us know if any adjustments are needed. Thank you.

**Comment by stnguyen90** [comment]
@EJM143, @SophieNguyen113 is still assigned on the issue. If there's no activity, I can reassign to you.

**Comment by EJM143** [comment]
@stnguyen90, Sophie and I are part of the same team. This issue was assigned to our team and as a team, I am submitting our new PR. If the feature assignee and the PR submitter need to be the same person, please reassign the feature to me. Thank you!

**Comment by DH-555** [comment]
@EJM143 I assigned you to the issue

**Comment by DH-555** [comment]
@EJM143 Are you working on this?

**Comment by magic-peach** [comment]
Hey, I’m participating in Hacktoberfest! Can I work on this issue?

**Comment by alokdangre** [comment]
@DH-555 can you assign it to me, since no one is working so, i want to work on this

**Comment by eldadfux** [comment]
We have decided not to move forward with this feature as it attempts to force a specific user journey on all use cases. In some cases blocking a user might be temporary and not require deleting all sessions. The API does allow to handle this by calling the method to delete all sessions in addition to blocking a user so this doesn't unlock an option that wasn't possible before.
